account_invoice_change_currency/tests/test_account_invoice_change_currency.py

```diff
@@ -0,0 +1,334 @@
+        after_curr = self.env.ref("base.USD")
+        inv.write({"currency_id": after_curr.id})
+        inv._onchange_currency_change_rate()
+        inv.action_account_change_currency()
+        expected_value = before_curr._convert(
+            before_amount, after_curr, inv.company_id, fields.Date.today()
+        )
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, 0),
+            0,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+
+    def test_change_validated_invoice_currency(self):
+        inv = self.create_simple_invoice(fields.Date.today())
+        before_amount = inv.amount_total
+        inv.action_post()
+        # Make sure that we can not change the currency after validated:
+        inv.write({"currency_id": self.env.ref("base.USD").id})
+        inv._onchange_currency_change_rate()
+        inv.action_account_change_currency()
+        self.assertEqual(
+            inv.amount_total,
+            before_amount,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+
+    def test_create_invoice_update_currency(self):
+        inv = self.create_simple_invoice()
+        before_amount = inv.amount_total
+        inv.action_account_change_currency()
+        self.assertEqual(
+            inv.amount_total,
+            before_amount,
+            "Amount must remain the same, because no currency changes",
+        )
+
+    def test_custom_rate_update_currency(self):
+        inv = self.create_simple_invoice(fields.Date.today())
+        before_amount = inv.amount_total
+        after_curr = self.env.ref("base.USD")
+        custom_rate = 1.13208
+        inv.write({"currency_id": after_curr.id, "custom_rate": custom_rate})
+        inv._onchange_currency_change_rate()
+        inv.write({"custom_rate": custom_rate})
+        inv.action_account_change_currency()
+        expected_value = before_amount * custom_rate
+        # TODO: Check float comparation, 12013.64 vs 12013.632959999999
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, self.precision),
+            1,
+            "Total amount of invoice does not equal to expected value!!!",
+        )
+
+    def test_custom_rate_zero_update_currency(self):
+        inv = self.create_simple_invoice()
+        before_amount = inv.amount_total
+        before_curr = inv.currency_id
+        custom_rate = 0.0
+        usd = self.env.ref("base.USD")
+        eur = self.env.ref("base.EUR")
+        inv.write({"currency_id": usd.id, "custom_rate": custom_rate})
+        inv._onchange_currency_change_rate()
+        inv.write({"custom_rate": custom_rate})
+        inv.action_account_change_currency()
+        expected_value = before_curr._convert(
+            before_amount, usd, inv.company_id, fields.Date.today()
+        )
+        self.assertEqual(
+            float_compare(inv.amount_total, expected_value, 0),
```

Comment:
Why?

Reply:
This code makes a comparison between 2004.64 vs 2004.67